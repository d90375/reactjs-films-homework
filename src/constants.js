export const IMAGE_NOT_FOUND_URL =
  "https://www.themoviedb.org/assets/2/v4/glyphicons/basic/glyphicons-basic-38-picture-grey-c2ebdbb057f2a7614185931650f8cee23fa137b93812ccb132b9df511df1cfac.svg";
export const NUMBER_OF_CARDS = 16;
export const KEY = "?api_key=6da466b0eb7061f0a6aba0e23f44d47d";
export const API = "https://api.themoviedb.org/3/";
export const LANG = "&language=en-US";
export const QUERY = "&query=";
export const PAGE = "&page=1";
export const ADULT = "&include_adult=false";
export const REGION = "US";
