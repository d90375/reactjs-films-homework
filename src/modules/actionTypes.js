export const FETCH_MOVIES_PENDING = "FETCH_MOVIES_PENDING";
export const FETCH_MOVIES_FULFILLED = "FETCH_MOVIES_FULFILLED";
export const FETCH_MOVIES_REJECTED = "FETCH_MOVIES_REJECTED";
export const FETCH_MOVIES_GENRES = "FETCH_MOVIES_GENRES";
export const FETCH_MOVIES_GENRES_BY_ID = "FETCH_MOVIES_GENRES_BY_ID";

export const FETCH_CURRENT_MOVIE_PENDING = "FETCH_CURRENT_MOVIE_PENDING";
export const FETCH_CURRENT_MOVIE_REJECTED = "FETCH_CURRENT_MOVIE_REJECTED";
export const FETCH_CURRENT_MOVIE_FULFILLED = "FETCH_CURRENT_MOVIE_FULFILLED";

export const FETCH_TRAILER_PENDING = "FETCH_TRAILER_PENDING";
export const FETCH_TRAILER_REJECTED = "FETCH_TRAILER_REJECTED";
export const FETCH_TRAILER_FULFILLED = "FETCH_TRAILER_FULFILLED";
export const FETCH_TRAILER_REMOVE = "FETCH_TRAILER_REMOVE";
