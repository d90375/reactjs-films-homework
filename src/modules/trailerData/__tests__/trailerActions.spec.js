import React from "react";
import { Provider } from "react-redux";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { renderHook } from "@testing-library/react-hooks";
import useTrailerActions from "../trailerActions";
import { API, KEY, LANG } from "../../../constants";
import { FETCH_TRAILER_FULFILLED, FETCH_TRAILER_PENDING, FETCH_TRAILER_REJECTED } from "../../actionTypes";

describe("modules > trailer > useTrailer", () => {
  const id = "724989";

  const cfgTrailerUrl = `${API}movie/${id}/videos${KEY}${LANG}`;

  const mockStore = configureStore([thunk]);

  const store = mockStore({
    trailer: {
      isLoadingTrailer: false,
      isFulfilledTrailer: false,
      hasErrorTrailer: false
    }
  });

  const setUp = (props) => {
    const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;
    const { result } = renderHook(() => useTrailerActions(props), {
      wrapper
    });

    return result;
  };

  it("returns function", () => {
    const result = setUp(724989);

    expect(result.current.handleShowTrailer).toBeInstanceOf(Function);
  });

  describe("handleShowTrailer", () => {
    const mockAxios = new MockAdapter(axios);

    const mockNetworkError = () => {
      mockAxios.onGet(cfgTrailerUrl).networkError();
    };

    const mock404 = () => {
      mockAxios.onGet(cfgTrailerUrl).reply(404);
    };

    const mockTimeout = () => {
      mockAxios.onGet(cfgTrailerUrl).timeout();
    };

    afterEach(() => {
      mockAxios.resetHandlers();
      store.clearActions();
    });

    it(`it handles successful API query`, async () => {
      const result = setUp(724989);

      const response = {
        id: 724989,
        results: [
          {
            key: "7Y6-w5Psupw"
          }
        ]
      };

      mockAxios.onGet(cfgTrailerUrl).reply(200, response);

      await result.current.handleShowTrailer();

      expect(store.getActions()[0]).toEqual({
        type: FETCH_TRAILER_PENDING
      });
      expect(store.getActions()[1].type).toEqual(FETCH_TRAILER_FULFILLED);
      expect(store.getActions()[1].payload.data).toEqual(response);
    });

    it(`it handle reset to default state`, async () => {
      const result = setUp(724989);

      const response = {
        id: 0,
        results: [
          {
            key: "qwe"
          }
        ]
      };

      mockAxios.onGet(cfgTrailerUrl).reply(200, response);

      await result.current.handleShowTrailer();
      await result.current.handleRemoveVideoFrame();

      expect(store.getActions()[1].payload.data).toEqual(response);
    });

    it.each([[mockNetworkError], [mock404], [mockTimeout]])(`it handles API fetching errors`, async (response) => {
      let hasErrorThrown;

      const result = setUp(724989);

      response();

      try {
        await result.current.handleShowTrailer();
      } catch {
        hasErrorThrown = true;
      } finally {
        expect(store.getActions()[0]).toEqual({
          type: FETCH_TRAILER_PENDING
        });
        expect(store.getActions()[1].type).toEqual(FETCH_TRAILER_REJECTED);
        expect(store.getActions()[1].payload).toBeInstanceOf(Error);
        expect(store.getActions()[1].payload).toMatchSnapshot();
        expect(hasErrorThrown).toBe(true);
      }
    });
  });
});
