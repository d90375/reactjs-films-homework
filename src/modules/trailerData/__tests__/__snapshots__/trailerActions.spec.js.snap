// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules > trailer > useTrailer handleShowTrailer it handles API fetching errors 1`] = `[Error: Network Error]`;

exports[`modules > trailer > useTrailer handleShowTrailer it handles API fetching errors 2`] = `[Error: Request failed with status code 404]`;

exports[`modules > trailer > useTrailer handleShowTrailer it handles API fetching errors 3`] = `[Error: timeout of 0ms exceeded]`;
